From 97964e07ac7e7043fea2dc57e8c819ce66578ac0 Mon Sep 17 00:00:00 2001
From: Jakub Adam <jakub.adam@ktknet.cz>
Date: Thu, 9 Jun 2016 12:40:06 +0200
Subject: [PATCH] Fix sipe-collab 14.04 build

---
 src/purple/purple-media.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/purple/purple-media.c b/src/purple/purple-media.c
index 1973d7b..7fdbb82 100644
--- a/src/purple/purple-media.c
+++ b/src/purple/purple-media.c
@@ -64,6 +64,9 @@
 #include <gst/rtp/gstrtcpbuffer.h>
 #include <farstream/fs-session.h>
 
+#undef GST_MAP_INFO_INIT
+#define GST_MAP_INFO_INIT { NULL, 0, NULL, 0, 0, {0, }, {0, }}
+
 struct sipe_backend_media {
 	PurpleMedia *m;
 	/**
-- 
2.8.1

