--- a/agent/agent.h 
+++ b/agent/agent.h 
@@ -213,6 +213,7 @@
  * @NICE_COMPATIBILITY_MSN: Use compatibility for MSN Messenger specs
  * @NICE_COMPATIBILITY_WLM2009: Use compatibility with Windows Live Messenger
  * 2009 and Microsoft Office Communicator 2007 R2
+ * @NICE_COMPATIBILITY_OC2007: Use compatibility with Microsoft Office Communicator 2007
  * @NICE_COMPATIBILITY_DRAFT19: Use compatibility for ICE Draft 19 specs
  * @NICE_COMPATIBILITY_LAST: Dummy last compatibility mode
  *
@@ -229,8 +230,9 @@
   NICE_COMPATIBILITY_GOOGLE,
   NICE_COMPATIBILITY_MSN,
   NICE_COMPATIBILITY_WLM2009,
+  NICE_COMPATIBILITY_OC2007,
   NICE_COMPATIBILITY_DRAFT19 = NICE_COMPATIBILITY_RFC5245,
-  NICE_COMPATIBILITY_LAST = NICE_COMPATIBILITY_WLM2009,
+  NICE_COMPATIBILITY_LAST = NICE_COMPATIBILITY_OC2007,
 } NiceCompatibility;
 
 /**

--- a/agent/agent.c 
+++ b/agent/agent.c 
@@ -151,6 +151,8 @@
       STUN_USAGE_ICE_COMPATIBILITY_MSN :
       agent->compatibility == NICE_COMPATIBILITY_WLM2009 ?
       STUN_USAGE_ICE_COMPATIBILITY_WLM2009 :
+      agent->compatibility == NICE_COMPATIBILITY_OC2007 ?
+      STUN_USAGE_ICE_COMPATIBILITY_MSN :
       STUN_USAGE_ICE_COMPATIBILITY_RFC5245;
 }
 
@@ -163,6 +165,8 @@
       agent->compatibility == NICE_COMPATIBILITY_MSN ?
       STUN_USAGE_TURN_COMPATIBILITY_MSN :
       agent->compatibility == NICE_COMPATIBILITY_WLM2009 ?
+      STUN_USAGE_TURN_COMPATIBILITY_MSN :
+      agent->compatibility == NICE_COMPATIBILITY_OC2007 ?
       STUN_USAGE_TURN_COMPATIBILITY_MSN : STUN_USAGE_TURN_COMPATIBILITY_DRAFT9;
 }
 
@@ -174,6 +178,8 @@
       agent->compatibility == NICE_COMPATIBILITY_MSN ?
       NICE_TURN_SOCKET_COMPATIBILITY_MSN :
       agent->compatibility == NICE_COMPATIBILITY_WLM2009 ?
+      NICE_TURN_SOCKET_COMPATIBILITY_MSN :
+      agent->compatibility == NICE_COMPATIBILITY_OC2007 ?
       NICE_TURN_SOCKET_COMPATIBILITY_MSN :
       NICE_TURN_SOCKET_COMPATIBILITY_DRAFT9;
 }
@@ -832,7 +838,8 @@
             STUN_COMPATIBILITY_RFC3489,
             STUN_AGENT_USAGE_SHORT_TERM_CREDENTIALS |
             STUN_AGENT_USAGE_IGNORE_CREDENTIALS);
-      } else if (agent->compatibility == NICE_COMPATIBILITY_MSN) {
+      } else if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+		 agent->compatibility == NICE_COMPATIBILITY_OC2007) {
         stun_agent_init (&agent->stun_agent, STUN_ALL_KNOWN_ATTRIBUTES,
             STUN_COMPATIBILITY_RFC3489,
             STUN_AGENT_USAGE_SHORT_TERM_CREDENTIALS |
@@ -1439,7 +1446,8 @@
             STUN_AGENT_USAGE_SHORT_TERM_CREDENTIALS |
             STUN_AGENT_USAGE_IGNORE_CREDENTIALS);
       } else if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
-                 agent->compatibility == NICE_COMPATIBILITY_WLM2009) {
+                 agent->compatibility == NICE_COMPATIBILITY_WLM2009 ||
+                 agent->compatibility == NICE_COMPATIBILITY_OC2007) {
         stun_agent_init (&cdisco->stun_agent, STUN_ALL_KNOWN_ATTRIBUTES,
             STUN_COMPATIBILITY_RFC3489,
             STUN_AGENT_USAGE_SHORT_TERM_CREDENTIALS);

--- a/agent/conncheck.c 
+++ b/agent/conncheck.c 
@@ -976,7 +976,8 @@
           NiceCandidate *remote_candidate = NULL;
 
           if (agent->compatibility == NICE_COMPATIBILITY_GOOGLE ||
-              agent->compatibility == NICE_COMPATIBILITY_MSN) {
+              agent->compatibility == NICE_COMPATIBILITY_MSN ||
+              agent->compatibility == NICE_COMPATIBILITY_OC2007) {
             /* We need to find which local candidate was used */
             uint8_t uname[NICE_STREAM_MAX_UNAME];
             guint uname_len;
@@ -1480,7 +1481,8 @@
       len += remote_len;
       memcpy (dest + len, local, local_len);
       len += local_len;
-    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN) {
+    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+	       agent->compatibility == NICE_COMPATIBILITY_OC2007) {
       gchar component_str[10];
       guchar *local_decoded = NULL;
       guchar *remote_decoded = NULL;
@@ -1636,7 +1638,8 @@
   size_t buffer_len;
   unsigned int timeout;
 
-  if (agent->compatibility == NICE_COMPATIBILITY_MSN) {
+  if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+      agent->compatibility == NICE_COMPATIBILITY_OC2007) {
     password = g_base64_decode ((gchar *) password, &password_len);
   }
 
@@ -1668,7 +1671,8 @@
 
     nice_debug ("Agent %p: conncheck created %d - %p", agent, buffer_len, pair->stun_message.buffer);
 
-    if (agent->compatibility == NICE_COMPATIBILITY_MSN) {
+    if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+        agent->compatibility == NICE_COMPATIBILITY_OC2007) {
       g_free (password);
     }
 
@@ -2551,7 +2555,17 @@
   gchar *ufrag = NULL;
   gsize ufrag_len;
 
-  for (i = data->component->local_candidates; i; i = i->next) {
+  gboolean msn_msoc_nice_compatibility =
+      data->agent->compatibility == NICE_COMPATIBILITY_MSN ||
+      data->agent->compatibility == NICE_COMPATIBILITY_OC2007;
+
+  if (data->agent->compatibility == NICE_COMPATIBILITY_OC2007 &&
+      stun_message_get_class (message) == STUN_RESPONSE)
+    i = data->component->remote_candidates;
+  else
+    i = data->component->local_candidates;
+
+  for (; i; i = i->next) {
     NiceCandidate *cand = i->data;
 
     ufrag = NULL;
@@ -2561,7 +2575,7 @@
       ufrag = data->stream->local_ufrag;
     ufrag_len = ufrag? strlen (ufrag) : 0;
 
-    if (ufrag && data->agent->compatibility == NICE_COMPATIBILITY_MSN)
+    if (ufrag && msn_msoc_nice_compatibility)
       ufrag = (gchar *)g_base64_decode (ufrag, &ufrag_len);
 
     if (ufrag == NULL)
@@ -2586,20 +2600,20 @@
         *password = (uint8_t *) pass;
         *password_len = strlen (pass);
 
-        if (data->agent->compatibility == NICE_COMPATIBILITY_MSN) {
+        if (msn_msoc_nice_compatibility) {
           data->password = g_base64_decode (pass, password_len);
           *password = data->password;
         }
       }
 
-      if (data->agent->compatibility == NICE_COMPATIBILITY_MSN)
+      if (msn_msoc_nice_compatibility)
         g_free (ufrag);
 
       stun_debug ("Found valid username, returning password: '%s'\n", *password);
       return TRUE;
     }
 
-    if (data->agent->compatibility == NICE_COMPATIBILITY_MSN)
+    if (msn_msoc_nice_compatibility)
       g_free (ufrag);
   }
 
@@ -2722,7 +2736,8 @@
     if (len == 0)
       return FALSE;
 
-    if (agent->compatibility != NICE_COMPATIBILITY_MSN) {
+    if (agent->compatibility != NICE_COMPATIBILITY_MSN &&
+        agent->compatibility != NICE_COMPATIBILITY_OC2007) {
       nice_socket_send (socket, from, rbuf_len, (const gchar*)rbuf);
     }
     return TRUE;
@@ -2734,7 +2749,8 @@
     if (stun_agent_init_error (&agent->stun_agent, &msg, rbuf, rbuf_len,
             &req, STUN_ERROR_UNAUTHORIZED)) {
       rbuf_len = stun_agent_finish_message (&agent->stun_agent, &msg, NULL, 0);
-      if (rbuf_len > 0 && agent->compatibility != NICE_COMPATIBILITY_MSN)
+      if (rbuf_len > 0 && agent->compatibility != NICE_COMPATIBILITY_MSN &&
+          agent->compatibility != NICE_COMPATIBILITY_OC2007)
         nice_socket_send (socket, from, rbuf_len, (const gchar*)rbuf);
     }
     return TRUE;
@@ -2744,7 +2760,8 @@
     if (stun_agent_init_error (&agent->stun_agent, &msg, rbuf, rbuf_len,
             &req, STUN_ERROR_BAD_REQUEST)) {
       rbuf_len = stun_agent_finish_message (&agent->stun_agent, &msg, NULL, 0);
-      if (rbuf_len > 0 && agent->compatibility != NICE_COMPATIBILITY_MSN)
+      if (rbuf_len > 0 && agent->compatibility != NICE_COMPATIBILITY_MSN &&
+	  agent->compatibility != NICE_COMPATIBILITY_OC2007)
         nice_socket_send (socket, from, rbuf_len, (const gchar*)rbuf);
     }
     return TRUE;
@@ -2762,7 +2779,8 @@
   }
 
   if (agent->compatibility == NICE_COMPATIBILITY_GOOGLE ||
-      agent->compatibility == NICE_COMPATIBILITY_MSN) {
+      agent->compatibility == NICE_COMPATIBILITY_MSN ||
+      agent->compatibility == NICE_COMPATIBILITY_OC2007) {
     /* We need to find which local candidate was used */
     for (i = component->remote_candidates;
          i != NULL && remote_candidate2 == NULL; i = i->next) {
@@ -2772,7 +2790,8 @@
         NiceCandidate *lcand = j->data;
 
         /* If we receive a response, then the username is local:remote */
-        if (agent->compatibility != NICE_COMPATIBILITY_MSN) {
+        if (agent->compatibility != NICE_COMPATIBILITY_MSN &&
+            agent->compatibility != NICE_COMPATIBILITY_OC2007) {
           if (stun_message_get_class (&req) == STUN_REQUEST ||
               stun_message_get_class (&req) == STUN_INDICATION) {
             inbound = TRUE;
@@ -2824,16 +2843,23 @@
 
 
   if (stun_message_get_class (&req) == STUN_REQUEST) {
-    if (agent->compatibility == NICE_COMPATIBILITY_MSN) {
+    if (   agent->compatibility == NICE_COMPATIBILITY_MSN
+        || agent->compatibility == NICE_COMPATIBILITY_OC2007) {
       if (local_candidate && remote_candidate2) {
-        username = (uint8_t *) stun_message_find (&req,
-            STUN_ATTRIBUTE_USERNAME, &username_len);
-        uname_len = priv_create_username (agent, stream,
-            component->id,  remote_candidate2, local_candidate,
-            uname, sizeof (uname), FALSE);
-        memcpy (username, uname, username_len);
-        req.key = g_base64_decode ((gchar *) remote_candidate2->password,
-            &req.key_len);
+	if (agent->compatibility == NICE_COMPATIBILITY_MSN) {
+          username = (uint8_t *) stun_message_find (&req,
+	  STUN_ATTRIBUTE_USERNAME, &username_len);
+	  uname_len = priv_create_username (agent, stream,
+              component->id,  remote_candidate2, local_candidate,
+	      uname, sizeof (uname), FALSE);
+	  memcpy (username, uname, username_len);
+
+	  req.key = g_base64_decode ((gchar *) remote_candidate2->password,
+              &req.key_len);
+	} else if (agent->compatibility == NICE_COMPATIBILITY_OC2007) {
+          req.key = g_base64_decode ((gchar *) local_candidate->password,
+              &req.key_len);
+	}
       } else {
         nice_debug ("Agent %p : received MSN incoming check from unknown remote candidate. "
             "Ignoring request", agent);
@@ -2847,7 +2873,8 @@
         &control, agent->tie_breaker,
         agent_to_ice_compatibility (agent));
 
-    if (agent->compatibility == NICE_COMPATIBILITY_MSN) {
+    if (   agent->compatibility == NICE_COMPATIBILITY_MSN
+        || agent->compatibility == NICE_COMPATIBILITY_OC2007) {
       g_free (req.key);
     }
 
@@ -2863,7 +2890,8 @@
 
       if (agent->controlling_mode ||
           agent->compatibility == NICE_COMPATIBILITY_GOOGLE ||
-          agent->compatibility == NICE_COMPATIBILITY_MSN)
+          agent->compatibility == NICE_COMPATIBILITY_MSN ||
+          agent->compatibility == NICE_COMPATIBILITY_OC2007)
         use_candidate = TRUE;
 
       if (stream->initial_binding_request_received != TRUE)

--- a/agent/discovery.c 
+++ b/agent/discovery.c 
@@ -414,7 +414,8 @@
     NiceCandidate *candidate)
 {
 
-  if (agent->compatibility == NICE_COMPATIBILITY_MSN) {
+  if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+      agent->compatibility == NICE_COMPATIBILITY_OC2007) {
     guchar username[32];
     guchar password[16];
 
@@ -475,7 +476,8 @@
     candidate->base_addr = *address;
     if (agent->compatibility == NICE_COMPATIBILITY_GOOGLE) {
       candidate->priority = nice_candidate_jingle_priority (candidate);
-    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN)  {
+    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+	       agent->compatibility == NICE_COMPATIBILITY_OC2007)  {
       candidate->priority = nice_candidate_msn_priority (candidate);
     } else {
       candidate->priority = nice_candidate_ice_priority (candidate);
@@ -556,7 +558,8 @@
   if (candidate) {
     if (agent->compatibility == NICE_COMPATIBILITY_GOOGLE) {
       candidate->priority = nice_candidate_jingle_priority (candidate);
-    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN)  {
+    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+               agent->compatibility == NICE_COMPATIBILITY_OC2007)  {
       candidate->priority = nice_candidate_msn_priority (candidate);
     } else {
       candidate->priority =  nice_candidate_ice_priority_full
@@ -616,7 +619,8 @@
   if (candidate) {
     if (agent->compatibility == NICE_COMPATIBILITY_GOOGLE) {
       candidate->priority = nice_candidate_jingle_priority (candidate);
-    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN)  {
+    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+               agent->compatibility == NICE_COMPATIBILITY_OC2007)  {
       candidate->priority = nice_candidate_msn_priority (candidate);
     } else {
       candidate->priority =  nice_candidate_ice_priority_full
@@ -707,7 +711,8 @@
     candidate->transport = NICE_CANDIDATE_TRANSPORT_UDP;
     if (agent->compatibility == NICE_COMPATIBILITY_GOOGLE) {
       candidate->priority = nice_candidate_jingle_priority (candidate);
-    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN)  {
+    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+	       agent->compatibility == NICE_COMPATIBILITY_OC2007)  {
       candidate->priority = nice_candidate_msn_priority (candidate);
     } else {
       candidate->priority = nice_candidate_ice_priority_full
@@ -721,7 +726,8 @@
 
     priv_assign_foundation (agent, candidate);
 
-    if (agent->compatibility == NICE_COMPATIBILITY_MSN &&
+    if ((agent->compatibility == NICE_COMPATIBILITY_MSN ||
+         agent->compatibility == NICE_COMPATIBILITY_OC2007) &&
 	remote && local) {
       guchar *new_username = NULL;
       guchar *decoded_local = NULL;
@@ -805,7 +811,8 @@
       candidate->priority = priority;
     } else if (agent->compatibility == NICE_COMPATIBILITY_GOOGLE) {
       candidate->priority = nice_candidate_jingle_priority (candidate);
-    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN)  {
+    } else if (agent->compatibility == NICE_COMPATIBILITY_MSN ||
+	       agent->compatibility == NICE_COMPATIBILITY_OC2007)  {
       candidate->priority = nice_candidate_msn_priority (candidate);
     } else {
       candidate->priority = nice_candidate_ice_priority_full
@@ -817,7 +824,8 @@
 
     priv_assign_remote_foundation (agent, candidate);
 
-    if (agent->compatibility == NICE_COMPATIBILITY_MSN &&
+    if ((agent->compatibility == NICE_COMPATIBILITY_MSN ||
+         agent->compatibility == NICE_COMPATIBILITY_OC2007) &&
 	remote && local) {
       guchar *new_username = NULL;
       guchar *decoded_local = NULL;

