--- a/agent/agent.h 
+++ b/agent/agent.h 
@@ -212,7 +212,7 @@
  * @NICE_COMPATIBILITY_GOOGLE: Use compatibility for Google Talk specs
  * @NICE_COMPATIBILITY_MSN: Use compatibility for MSN Messenger specs
  * @NICE_COMPATIBILITY_WLM2009: Use compatibility with Windows Live Messenger
- * 2009
+ * 2009 and Microsoft Office Communicator 2007 R2
  * @NICE_COMPATIBILITY_DRAFT19: Use compatibility for ICE Draft 19 specs
  * @NICE_COMPATIBILITY_LAST: Dummy last compatibility mode
  *

--- a/agent/agent.c 
+++ b/agent/agent.c 
@@ -149,6 +149,8 @@
       STUN_USAGE_ICE_COMPATIBILITY_GOOGLE :
       agent->compatibility == NICE_COMPATIBILITY_MSN ?
       STUN_USAGE_ICE_COMPATIBILITY_MSN :
+      agent->compatibility == NICE_COMPATIBILITY_WLM2009 ?
+      STUN_USAGE_ICE_COMPATIBILITY_WLM2009 :
       STUN_USAGE_ICE_COMPATIBILITY_RFC5245;
 }
 
--- a/agent/conncheck.c 
+++ b/agent/conncheck.c 
@@ -599,6 +599,7 @@
                 uname, uname_len, password, password_len,
                 agent->controlling_mode, agent->controlling_mode, priority,
                 agent->tie_breaker,
+                NULL,
                 agent_to_ice_compatibility (agent));
 
             nice_debug ("Agent %p: conncheck created %d - %p",
@@ -1662,6 +1663,7 @@
         uname, uname_len, password, password_len,
         cand_use, controlling, priority,
         agent->tie_breaker,
+        pair->foundation,
         agent_to_ice_compatibility (agent));
 
     nice_debug ("Agent %p: conncheck created %d - %p", agent, buffer_len, pair->stun_message.buffer);

--- a/stun/usages/ice.h 
+++ b/stun/usages/ice.h 
@@ -65,6 +65,8 @@
  * implementation of ICE
  * @STUN_USAGE_ICE_COMPATIBILITY_MSN: The ICE compatibility with MSN's
  * implementation of ICE
+ * @STUN_USAGE_ICE_COMPATIBILITY_WLM2009: The ICE compatibility with Windows
+ * Live Messenger and Microsoft Office Communicator 2007 R2 implementation of ICE
  * @STUN_USAGE_ICE_COMPATIBILITY_DRAFT19: The ICE compatibility with draft 19
  *
  * This enum defines which compatibility modes this ICE usage can use
@@ -78,6 +80,7 @@
   STUN_USAGE_ICE_COMPATIBILITY_RFC5245,
   STUN_USAGE_ICE_COMPATIBILITY_GOOGLE,
   STUN_USAGE_ICE_COMPATIBILITY_MSN,
+  STUN_USAGE_ICE_COMPATIBILITY_WLM2009,
   STUN_USAGE_ICE_COMPATIBILITY_DRAFT19 = STUN_USAGE_ICE_COMPATIBILITY_RFC5245,
 } StunUsageIceCompatibility;
 
@@ -130,12 +133,16 @@
  * @priority: The value of the PRIORITY attribute
  * @tie: The value of the tie-breaker to put in the ICE-CONTROLLED or
  * ICE-CONTROLLING attribute
+ * @candidate_identifier: The foundation value to put in the
+ * CANDIDATE-IDENTIFIER attribute
  * @compatibility: The compatibility mode to use for building the conncheck
  * request
  *
  * Builds an ICE connectivity check STUN message.
  * If the compatibility is not #STUN_USAGE_ICE_COMPATIBILITY_RFC5245, the
  * @cand_use, @controlling, @priority and @tie arguments are not used.
+ * If the compatibility is not #STUN_USAGE_ICE_COMPATIBILITY_WLM2009, the
+ * @candidate_identifier argument is not used.
  * Returns: The length of the message built.
  */
 size_t
@@ -144,7 +151,8 @@
     const uint8_t *username, const size_t username_len,
     const uint8_t *password, const size_t password_len,
     bool cand_use, bool controlling, uint32_t priority,
-    uint64_t tie, StunUsageIceCompatibility compatibility);
+    uint64_t tie, const char *candidate_identifier,
+    StunUsageIceCompatibility compatibility);
 
 
 /**
 
--- a/stun/usages/ice.c 
+++ b/stun/usages/ice.c 
@@ -42,6 +42,7 @@
 
 #include <string.h>
 #include <assert.h>
+#include <stdlib.h>
 
 #ifdef _WIN32
 #include <winsock2.h>
@@ -64,13 +65,15 @@
     const uint8_t *username, const size_t username_len,
     const uint8_t *password, const size_t password_len,
     bool cand_use, bool controlling, uint32_t priority,
-    uint64_t tie, StunUsageIceCompatibility compatibility)
+    uint64_t tie, const char *candidate_identifier,
+    StunUsageIceCompatibility compatibility)
 {
   StunMessageReturn val;
 
   stun_agent_init_request (agent, msg, buffer, buffer_len, STUN_BINDING);
 
-  if (compatibility == STUN_USAGE_ICE_COMPATIBILITY_RFC5245) {
+  if (compatibility == STUN_USAGE_ICE_COMPATIBILITY_RFC5245 ||
+      compatibility == STUN_USAGE_ICE_COMPATIBILITY_WLM2009) {
     if (cand_use)
     {
       val = stun_message_append_flag (msg, STUN_ATTRIBUTE_USE_CANDIDATE);
@@ -95,6 +98,28 @@
         username, username_len);
     if (val != STUN_MESSAGE_RETURN_SUCCESS)
       return 0;
+  }
+
+  if (compatibility == STUN_USAGE_ICE_COMPATIBILITY_WLM2009) {
+    size_t identifier_len = strlen(candidate_identifier);
+    size_t buffer_len = identifier_len;
+    int modulo4 = identifier_len % 4;
+    uint8_t* buf;
+
+    if (modulo4)
+        buffer_len += 4 - modulo4;
+
+    buf = malloc(buffer_len);
+    memset(buf, 0, buffer_len);
+    memcpy(buf, candidate_identifier, identifier_len);
+
+    val = stun_message_append_bytes (msg, STUN_ATTRIBUTE_CANDIDATE_IDENTIFIER,
+            buf, buffer_len);
+
+    free(buf);
+
+    if (val != STUN_MESSAGE_RETURN_SUCCESS)
+		return 0;
   }
 
   return stun_agent_finish_message (agent, msg, password, password_len);

--- a/stun/stunmessage.h 
+++ b/stun/stunmessage.h 
@@ -218,6 +218,8 @@
  * defined by ICE draft 19
  * @STUN_ATTRIBUTE_ICE_CONTROLLING: The ICE-CONTROLLING optional attribute as
  * defined by ICE draft 19
+ * @STUN_ATTRIBUTE_CANDIDATE_IDENTIFIER: The CANDIDATE-IDENTIFIER attribute as
+ * defined by [MS-ICE2]
  *
  * Known STUN attribute types as defined by various RFCs and drafts
  */
@@ -288,7 +290,9 @@
   STUN_ATTRIBUTE_FINGERPRINT=0x8028,    /* RFC5389 */
   STUN_ATTRIBUTE_ICE_CONTROLLED=0x8029,    /* ICE-19 */
   STUN_ATTRIBUTE_ICE_CONTROLLING=0x802A,    /* ICE-19 */
-  /* 0x802B-0xFFFF */      /* reserved */
+  /* 0x802B-0x8053 */      /* reserved */
+  STUN_ATTRIBUTE_CANDIDATE_IDENTIFIER=0x8054    /* MS-ICE2 */
+  /* 0x8055-0xFFFF */      /* reserved */
 } StunAttribute;

